pinot-plugins/pinot-input-format/pinot-csv/src/main/java/org/apache/pinot/plugin/inputformat/csv/CSVRecordReader.java

```diff
@@ -202,158 +207,139 @@ public Map<String, Integer> getCSVHeaderMap() {
+    }
+
+    if (_bufferedReader != null) {
+      _bufferedReader.close();
     }
-    return _nextRecord != null;
   }
 
-  private void readNextLine(GenericRow reuse)
-      throws IOException {
-    while (_nextLine != null) {
-      try (Reader reader = new StringReader(_nextLine)) {
-        try (CSVParser csvParser = _format.parse(reader)) {
-          List<CSVRecord> csvRecords = csvParser.getRecords();
-          if (csvRecords != null && csvRecords.size() > 0) {
-            // There would be only one record as lines are read one after the other
-            CSVRecord record = csvRecords.get(0);
-            _recordExtractor.extract(record, reuse);
-            break;
-          } else {
-            // Can be thrown on: 1) Empty lines 2) Commented lines
-            throw new NoSuchElementException("Failed to find any records");
+  class LineIterator implements Iterator<CSVRecord> {
+    private String _nextLine;
+    private CSVRecord _current;
+
+    public LineIterator() {
+      init();
+    }
+
+    private void init() {
+      try {
+        if (_config.getHeader() != null) {
+          if (_config.isSkipHeader()) {
+            // When skip header config is set and header is supplied – skip the first line from the input file
+            _bufferedReader.readLine();
+            // turn off the property so that it doesn't interfere with further parsing
+            _format = _format.builder().setSkipHeaderRecord(false).build();
           }
-        } catch (Exception e) {
-          _skippedLinesCount++;
-          LOGGER.debug("Skipped input line: {} from file: {}", _nextLine, _dataFile, e);
-          // Find the next line that can be parsed
-          _nextLine = _bufferedReader.readLine();
+        } else {
+          // read the first line
+          String headerLine = _bufferedReader.readLine();
+          _headerMap = parseHeaderMapFromLine(_format, headerLine);
+          // If header isn't provided, the first line would be set as header and the 'skipHeader' property
+          // is set to false.
+          _format = _format.builder()
+              .setSkipHeaderRecord(false)
+              .setHeader(_headerMap.keySet().toArray(new String[0]))
+              .build();
         }
+        _nextLine = _bufferedReader.readLine();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
       }
     }
-    if (_nextLine != null) {
-      // Advance the pointer to the next line for future reading
-      _nextLine = _bufferedReader.readLine();
-    } else {
-      throw new RuntimeException("No more parseable lines. Line iterator reached end of file.");
-    }
-  }
 
-  private Map<String, Integer> parseLineAsHeader(String line)
-      throws IOException {
-    Map<String, Integer> headerMap;
-    try (StringReader stringReader = new StringReader(line)) {
-      try (CSVParser parser = _format.parse(stringReader)) {
-        headerMap = parser.getHeaderMap();
+    private CSVRecord getNextRecord() {
```

Comment:
Should we optimise this method to not initialise stringReader every time next is called
or is that out of scope?

Reply:
Yes. This is not ideal but I really wanted to keep the scope of the PR focussed to refactor. Let me tackle this in a different PR.